src/py/flwr/server/superlink/fleet/vce/backend/raybackend.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Ray backend for the Fleet API using the VCE."""
+
+import asyncio
+import pathlib
+from logging import INFO
+from typing import Callable, Dict, List, Tuple, Union
+
+from flwr.client.clientapp import ClientApp
+from flwr.common.context import Context
+from flwr.common.logger import log
+from flwr.common.message import Message
+from flwr.simulation.ray_transport.ray_actor import (
+    BasicActorPool,
+    ClientAppActor,
+    init_ray,
+)
+
+from .backend import Backend, BackendConfig
+
+ClienteResourcesDict = Dict[str, Union[int, float]]
+
+
+class RayBackend(Backend):
+    """A backend that submits jobs to a `BasicActorPool`."""
+
+    def __init__(
+        self,
+        backend_config: BackendConfig,
+        work_dir: str,
+    ) -> None:
+        """Prepare RayBackend by initialising Ray and creating the ActorPool."""
+        log(INFO, "Initialising: %s", self.__class__.__name__)
+        log(INFO, "Backend config: %s", backend_config)
+
+        # Init ray and append working dir if needed
+        runtime_env = (
+            self._configure_runtime_env(work_dir=work_dir) if work_dir else None
+        )
+        init_ray(runtime_env=runtime_env)
+
+        # Validate client resources
+        self.client_resources_key = "client_resources"
+
+        # Create actor pool
+        actor_kwargs = backend_config.get("actor_kwargs", {})
+        client_resources = self._validate_client_resources(config=backend_config)
+        self.pool = BasicActorPool(
+            actor_type=ClientAppActor,
+            client_resources=client_resources,
+            actor_kwargs=actor_kwargs,
+        )
+
+    def _configure_runtime_env(self, work_dir: str) -> Dict[str, Union[str, List[str]]]:
+        """Return list of files/subdirectories to exclude relateive to work_dir.
+
+        Without this, Ray will push everything to the Ray Cluster.
+        """
+        runtime_env: Dict[str, Union[str, List[str]]] = {"working_dir": work_dir}
+
+        if runtime_env:
```

Comment:
Why do we need the `if` here? Given that the key `"working_dir"` will always be present, `if runtime_env:` will always return `True`, no?

Reply:
yes, no needed. fixed.